### PR TITLE
Custom font support

### PR DESCRIPTION
**Pull request by Gillibald**
- What does the pull request do?
It introduces support for loading fonts from embedded resources. 
- What is the current behavior?
Only system installed fonts are supported.
- What is the updated/expected behavior with this PR?
You can now define a FontFamily with resource syntax 
- How was the solution implemented (if it's not obvious)?
To speed up loading times and save memory resources are cached. Items of that cache are idetified by a FontFamilyKey that holds a location and optional filename. If you define several FontFamilies with same location and filename it only loads resources once.

To define a FontFamily you can use the commonly used resm scheme. If you point to a resource folder all containing fonts are automaticly discovered. It also introduces filename globbing or so called wildcard support of filenames. That way you can define a FontFamily and automaticly lookup all typefaces in one line.

You can also define fallbacks for your FontFamily. If a font isnt installed on the system, the logic tries to use the fallback values you defined for your FontFamily to render the text.

Some examples of syntax:

Installed font
`
FontFamily="Arial"
`
Installed font with fallbacks
`
FontFamily="Arial, Times New Roman, Comic Sans MS"
`
All font files in a folder
`
FontFamily="resm:DefaultNamespace.Fonts#MyFont"
`
Single font file
`
FontFamily="resm:DefaultNamespace.Fonts.MyFont.ttf#MyFont"
`
All font files in a folder that match expression
`
FontFamily="resm:DefaultNamespace.Fonts.MyFont*.ttf#MyFont"
`

If your font resources are not in your main assembly you have to define the assembly attribute in your uri strings.

To use specific font weights or styles make sure your font is supporting them. If you want to have support for multiple typefaces you have to add them by folder. Your resources would look like this:

```
DefaultNamespace.Fonts.MyFont.MyFont_Regular.ttf
DefaultNamespace.Fonts.MyFont.MyFont_Bold.ttf
DefaultNamespace.Fonts.MyFont.MyFont_Italic.ttf
DefaultNamespace.Fonts.MyFont.MyFont_Italic_Bold.ttf
etc.
```
Then you can use them like so:

```
FontFamily="resm:DefaultNamespace.Fonts.MyFont#MyFont"
FontFamily="resm:DefaultNamespace.Fonts.MyFont#MyFont" FontWeight="Bold"
FontFamily="resm:DefaultNamespace.Fonts.MyFont#MyFont" FontStyle="Italic"
FontFamily="resm:DefaultNamespace.Fonts.MyFont#MyFont" FontWeight="Bold" FontStyle="Italic"

```
I noticed some ui freaze running this under DotNetCore that is gone after everything is loaded. Have to figure out what is causing this.

With this approach we could also support loading fonts from filesystem. Let me know if thats a wanted feature or if we should introduce that later.

Checklist:

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?

See #1313 
See #1564 
See #810

### CODE REVIEW

**Comment by danwalmsley** [comment]
@Gillibald great work, i'll try and test this shortly.

Any ideas on how we can support font family lists:

(from msft docs)
fontFamilyNamesList
A string specifying multiple font family names, each separated by a comma (any whitespace following a comma is ignored). The first font family specified serves as the primary font family; subsequent font families serve as fallback families to be used in cases where the primary font family is unavailable or not applicable. For example, "Arial, Century Gothic" specifies Arial as the primary font family, with Century Gothic as the fallback font family.

i.e. 
```
FontFamily="Arial, Verdana, Comic Sans Ms"
```

**Comment by danwalmsley** [comment]
I think your PR should change the api of: TextFormatterFactory.CreateFormattedText to accept FontFamily instead of string for name of fontfamily?

After testing your PR I'm having to change that api in lots of places.

**Comment by Gillibald** [comment]
In addition to use a fallback for non present fonts we could use a fallback in case a specific typeface isn't present. For example if we set FontWeight to Bold and the main FontFamily has no face to represent that setting we could check if the fallback has one and use that. Not sure how to cache that properly but should be doable.

**Comment by danwalmsley** [comment]
I'm getting `System.MissingFieldException: 'Field not found: 'Avalonia.Controls.Primitives.TemplatedControl.FontFamilyProperty'.'` maybe I didn't build something right!

**Comment by Gillibald** [comment]
Will have a look at TextFormatterFactory.CreateFormattedText
This PR is tested with DirectD21 and Skia backend so it should just work. Even styles with FontFamily definitions should just work.

**Comment by danwalmsley** [comment]
Probably just add an extra overload that takes in fontfamily so we don't break existing api
TextFormatterFactory.CreateFormattedText

**Comment by Gillibald** [comment]
Are you sure that TextFormatterFactory.CreateFormattedText is part of this project?

**Comment by danwalmsley** [comment]
oops sorry! I was looking at avalonia edit code.

**Comment by danwalmsley** [comment]
Note: This PR will not be binary compatible due to change of FontFamily property type.

**Comment by Gillibald** [comment]
Need to add some render tests for DWrite and Skia implementation other than that this is feature complete and can be tested.

**Comment by danwalmsley** [comment]
@Gillibald this still doesn't work for me all fonts get resolved to Consolas or something similar.

image

Can you contact me on gitter so we can diagnose the issues?

File: src/Avalonia.Controls/TextBlock.cs [hunk 1]
```diff
@@ -227,7 +227,7 @@ public TextAlignment TextAlignment
         /// </summary>
         /// <param name="control">The control.</param>
         /// <returns>The font family.</returns>
-        public static string GetFontFamily(Control control)
+        public static FontFamily GetFontFamily(Control control)
```
Comment (danwalmsley) [inline src/Avalonia.Controls/TextBlock.cs]: You must also change the SetFontFamily method. 

```
public static void SetFontFamily(Control control, FontFamily value)
        {
            control.SetValue(FontFamilyProperty, value);
        }
```
currently it throws an exception when this is called.

File: src/Avalonia.Visuals/Media/FontFamily.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Avalonia.Media.Fonts;
+
+namespace Avalonia.Media
+{
+    public class FontFamily
+    {
+        internal static FontFamily Default = new FontFamily("Courier New");
```
Comment (danwalmsley) [inline src/Avalonia.Visuals/Media/FontFamily.cs]: Can we use the same font that is currently default? Some apps may be relying on this being consistent.

This should be a font or series of fonts available cross platform, perhaps we can see what WPF default it, and have font fallbacks to similar fonts for Linux, mac and ios / android.

**Review by danwalmsley — CHANGES_REQUESTED** [review]

**Comment by danwalmsley** [comment]
I also get `System.Exception: 'resm:AvalonStudio.Assets.SourceSansPro-Regular.ttf;assembly=AvalonStudio#Source Sans Pro is not a valid value for Double.'`

If I try to use a font from a different assembly.

**Comment by danwalmsley** [comment]
@Gillibald I wonder is it possible to have it so once a font has been loaded say declared as a resource:
like:
`<FontFamily x:Key="EditorFont">resm:AvalonStudio.Assets.SourceCodePro-Regular.ttf#Source Code Pro</FontFamily>`

It could be then accessed like FontFamily="Source Code Pro" 

where it gets located in the cache?

File: src/Avalonia.Visuals/Media/FontFamily.cs (hunk 2 above)
Comment (Gillibald) [reply src/Avalonia.Visuals/Media/FontFamily.cs]: I just used what already was the default just defined it as a static property. This should be used everywhere in our code base when the default is needed.

**Comment by danwalmsley** [comment]
Also if I set a FontWeight i.e. Bold it resets the fontfamily to Courier New

**Comment by Gillibald** [comment]
I can tests this PR against Avalon Studio if you want. Defining a custom font with full resource string and then reference it just by name at other places is not possible because you can't identify it properly. You still can define your FontFamily in a ResourceDictionary and reference that in your controls.

I can't emaging how you would get an Exception related to a double value.

**Comment by Gillibald** [comment]
If you font can't represent the font weight the default is used. That an expected result. You need to load the correct font for your FontWeight and Style

**Comment by danwalmsley** [comment]
I'm posting a repro for control catalog, that shows the fonts wont seem to load from separate dll to the main executable.

Repros available here:
https://github.com/AvaloniaUI/Avalonia/tree/repros/custom-fonts-pr

Just load the TextBox page, font isn't resolved, the fonts are embedded in ControlCatalog/Assets/...

It will crash immediately pending the fix to TextBlock.SetFontFamily api. this is an easy change.

**Comment by Gillibald** [comment]
I have tested your repro and could run everything with the fix to TextBlock.SetFontFamily. But i had to fix your resource strings a bit. The Default assembly in your case wasn't ControlCatalog so it could't resolve that. I had to change it like so:

`FontFamily="resm:ControlCatalog.Assets.SourceSansPro-Regular.ttf?assembly=ControlCatalog#Source Sans Pro" 
`

The default assembly is always the entry assembly and thats for example ControlCatalog.Desktop

File: src/Avalonia.Visuals/Media/Typeface.cs [hunk 3]
```diff
@@ -7,18 +7,8 @@ namespace Avalonia.Media
...
-        /// <param name="fontFamilyName">The name of the font family.</param>
-        /// <param name="fontSize">The font size, in DIPs.</param>
-        /// <param name="style">The font style.</param>
-        /// <param name="weight">The font weight.</param>
-        public Typeface(
-            string fontFamilyName,
-            double fontSize,
-            FontStyle style = FontStyle.Normal,
-            FontWeight weight = FontWeight.Normal)
+        public Typeface(FontFamily fontFamily, double fontSize = 12, FontStyle style = FontStyle.Normal,
```
Comment (danwalmsley) [inline src/Avalonia.Visuals/Media/Typeface.cs]: Need to add documentation to this method.

**Comment by danwalmsley** [comment]
Ok im going to try folder loading now.

**Comment by danwalmsley** [comment]
This is looking really great btw, hopefully we can get it merged soon.

**Comment by Gillibald** [comment]
I have updated the description of this PR to explain how to use multiple typfaces. Have to write some docs.

**Comment by danwalmsley** [comment]
Ok pretty simple just use *. i.e.
`<FontFamily x:Key="InputFont">resm:ControlCatalog.Assets.SourceSansPro*.ttf?assembly=ControlCatalog#Source Sans Pro</FontFamily>`

**Comment by danwalmsley** [comment]
ok this is all working great, i'll see how things run on Linux in the morning.

File: src/Avalonia.Visuals/Media/Typeface.cs (hunk 3 above)
Comment (Gillibald) [reply src/Avalonia.Visuals/Media/Typeface.cs]: I think I should spend some time on comments in general. There is a lot room for improvements.

File: src/Avalonia.Visuals/Media/Fonts/FontFamilyLoader.cs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        }
+
+        /// <summary>
+        /// Searches for font resources at a given location and returns a quanity of found resources
+        /// </summary>
+        /// <param name="location"></param>
+        /// <returns></returns>
+        private static IEnumerable<FontResource> GetFontResourcesByLocation(Uri location)
+        {
+            var assembly = GetAssembly(location);
```
Comment (kekekeks) [inline src/Avalonia.Visuals/Media/Fonts/FontFamilyLoader.cs]: You shouldn't be accessing reflection API directly. That's what IAssetLoader is there for.

File: tests/Avalonia.RenderTests/Controls/BorderTests.cs [hunk 5]
```diff
@@ -204,7 +204,7 @@ public BorderTests()
                     {
                         Text = "Foo",
                         Background = Brushes.Red,
-                        FontFamily = "Segoe UI",
```
Comment (kekekeks) [inline tests/Avalonia.RenderTests/Controls/BorderTests.cs]: This is a breaking change. Please,  add an implicit convertion operator to the FontFamily class

Comment (Gillibald) [reply tests/Avalonia.RenderTests/Controls/BorderTests.cs]: Okay that's a good idea

File: src/Shared/PlatformSupport/AssetLoader.cs [hunk 6]
```diff
@@ -218,5 +216,5 @@ public AssemblyDescriptor(Assembly assembly)
             public Dictionary<string, IAssetDescriptor> Resources { get; }
             public string Name { get; }
         }
-    }
+    }
```
Comment (danwalmsley) [inline src/Shared/PlatformSupport/AssetLoader.cs]: Remove whitespace added here.

File: src/Windows/Avalonia.Direct2D1/Direct2D1Platform.cs [hunk 7]
```diff
@@ -117,7 +117,7 @@ public IBitmapImpl CreateBitmap(int width, int height)
         }
 
         public IFormattedTextImpl CreateFormattedText(
-            string text,
+            string text,
```
Comment (danwalmsley) [inline src/Windows/Avalonia.Direct2D1/Direct2D1Platform.cs]: remove whitespace change.

File: tests/Avalonia.RenderTests/Controls/BorderTests.cs (hunk 5 above)
Comment (Gillibald) [reply tests/Avalonia.RenderTests/Controls/BorderTests.cs]: Have implemented implicit conversion

**Comment by Gillibald** [comment]
I have added a GetAssets method to IAssetLoader and renamed resource to asset to keep the naming consistent. The changes to IAssetLoader should probably have its own PR. Will do that shortly and will revert the changes I made until that got merged.

**Comment by danwalmsley** [comment]
I will have a look soon. 😀

File: src/Shared/PlatformSupport/AssetLoader.cs [hunk 8]
```diff
@@ -83,7 +83,7 @@ public bool Exists(Uri uri, Uri baseUri = null)
         /// <exception cref="FileNotFoundException">
         /// The resource was not found.
         /// </exception>
-        public Tuple<Stream, Assembly> OpenAndGetAssembly(Uri uri, Uri baseUri = null)
+        public (Stream Stream, Assembly Assembly) OpenAndGetAssembly(Uri uri, Uri baseUri = null)
```
Comment (danwalmsley) [inline src/Shared/PlatformSupport/AssetLoader.cs]: Nit:
Could we use `camelCase` for tuple values. It would look odd when a tuple is deconstructed and you end up with locals that start with capitals.

File: src/Shared/PlatformSupport/AssetLoader.cs [hunk 9]
```diff
@@ -92,9 +92,18 @@ public bool Exists(Uri uri, Uri baseUri = null)
...
+
+        public IEnumerable<(string AbsolutePath, Assembly Assembly)> GetAssets(Uri location)
+        {
+            var assembly = GetAssembly(location);
+
+            return assembly?.Resources.Select(x => (x.Key, x.Value.Assembly)) ??
+                   Enumerable.Empty<(string AbsolutePath, Assembly Assembly)>();
         }
 
+
```
Comment (danwalmsley) [inline src/Shared/PlatformSupport/AssetLoader.cs]: unnecessary whitespace.

File: src/Windows/Avalonia.Direct2D1/Media/DWriteResourceFontFileStream.cs [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        void FontFileStream.ReadFileFragment(out IntPtr fragmentStart, long fileOffset, long fragmentSize, out IntPtr fragmentContext)
+        {
+            lock (this)
+            {
+                fragmentContext = IntPtr.Zero;
+
+                _stream.Position = fileOffset;
+
+                fragmentStart = _stream.PositionPointer;
+
```
Comment (danwalmsley) [inline src/Windows/Avalonia.Direct2D1/Media/DWriteResourceFontFileStream.cs]: Nit: unnecessary line space

**Review by danwalmsley — CHANGES_REQUESTED** [review]
Some minor nits

File: tests/Avalonia.RenderTests/Controls/BorderTests.cs (hunk 5 above)
Comment (danwalmsley) [reply tests/Avalonia.RenderTests/Controls/BorderTests.cs]: If there is no an implicit conversion (I couldn't where is it btw?) then this change is no longer necessary, please revert all the FontFamily = new FontFamily( "") changes back to FontFamily = "".

File: src/Avalonia.Visuals/Media/FontFamily.cs [hunk 11]
```diff
@@ -79,9 +79,23 @@ internal FamilyNameCollection FamilyNames
...
+            }
+
             return Name;
         }
 
+        /// <summary>
+        /// Implicit conversion of FontFamily to string
+        /// </summary>
+        /// <param name="fontFamily"></param>
+        public static implicit operator string(FontFamily fontFamily)
```
Comment (danwalmsley) [inline src/Avalonia.Visuals/Media/FontFamily.cs]: I think this should have been the other way around, i.e. string to FontFamily.

**Review by danwalmsley — CHANGES_REQUESTED** [review]

File: src/Shared/PlatformSupport/AssetLoader.cs (hunk 8 above)
Comment (Gillibald) [reply src/Shared/PlatformSupport/AssetLoader.cs]: I will change that.

File: src/Avalonia.Visuals/Media/FontFamily.cs (hunk 11 above)
Comment (Gillibald) [reply src/Avalonia.Visuals/Media/FontFamily.cs]: The FontFamily was defined as a string so if someone still has some code with FontFamily = "MyFont" it still works. The other way isn't really needed. There is already ToString

File: src/Avalonia.Visuals/Properties/AssemblyInfo.cs [hunk 12]
```diff
@@ -10,5 +10,7 @@
 [assembly: XmlnsDefinition("https://github.com/avaloniaui", "Avalonia.Media")]
 [assembly: XmlnsDefinition("https://github.com/avaloniaui", "Avalonia")]
 
+[assembly: InternalsVisibleTo("Avalonia.Direct2D1")]
```
Comment (grokys) [inline src/Avalonia.Visuals/Properties/AssemblyInfo.cs]: We shouldn't really be doing this - it should be possible for people to write their own rendering backends without us adding an `InternalsVisibleTo` attribute.

File: src/Avalonia.Visuals/Media/Fonts/FontAsset.cs [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace Avalonia.Media.Fonts
+{
+    internal class FontAsset
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="FontAsset"/> class.
+        /// </summary>
+        /// <param name="source">The source.</param>
+        public FontAsset(Uri source)
```
Comment (grokys) [inline src/Avalonia.Visuals/Media/Fonts/FontAsset.cs]: This class just contains a single `Uri` - why is this abstraction needed? Why not just pass a `Uri`?

**Review by grokys — CHANGES_REQUESTED** [review]
Looking really good! Just one question and one change I think we need.

**Comment by Gillibald** [comment]
Dont know if I am happy with IFontFamily. Pls let me know your thoughts.

File: tests/Avalonia.RenderTests/Controls/BorderTests.cs (hunk 5 above)
Comment (Gillibald) [reply tests/Avalonia.RenderTests/Controls/BorderTests.cs]: I am not sure if I should revert everything to use string. Implicit conversion is just needed for backwards compatibility.

Comment (danwalmsley) [reply tests/Avalonia.RenderTests/Controls/BorderTests.cs]: My though process was that then this PR only shows the necessary changes, however i'll let you make the best judgement on this one :)

**Review by danwalmsley — APPROVED** [review]
LGTM 👍

File: src/Avalonia.Visuals/Media/FontFamily.cs [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+        public FontFamily(string name, Uri source) : this(name)
+        {
+            Key = new FontFamilyKey(source);
+        }
+
+        public static FontFamily Default => new FontFamily("Courier New");
+
+        public string Name => FamilyNames.PrimaryFamilyName;
+      
+        FamilyNameCollection IFontFamily.FamilyNames => FamilyNames;
```
Comment (grokys) [inline src/Avalonia.Visuals/Media/FontFamily.cs]: I don't understand this: `IFontFamily` exposes `Name`, `FamilyNames` and `Key` which are implemented here as explicit interface methods and forwarded to `internal` properties? Why not just make these properties public? That would have the same effect, no?

And why is `IFontFamily` needed in that case? It doesn't look like there can be any other implementation of it, as the `TextBlock.FontFamily` property is of type `FontFamily`. What use-case would having another implementation of `FontFamily` solve?

**Review by grokys — CHANGES_REQUESTED** [review]
Hey, I don't understand the `IFontFamily` changes - I know we talked about this in gitter and I thought that it was supposed to hide some implementation details, but I can't see how it's doing that here.

File: src/Avalonia.Visuals/Media/FontFamily.cs (hunk 14 above)
Comment (Gillibald) [reply src/Avalonia.Visuals/Media/FontFamily.cs]: Have removed the interface. It was ment to hide FamilyNames and Key for 3rd Party. The internal properties just made it easier to access them without casting.

File: samples/ControlCatalog/ControlCatalog.csproj [hunk 15]
```diff
@@ -11,6 +11,18 @@
     </EmbeddedResource>
     <EmbeddedResource Include="Assets\*" />
   </ItemGroup>
+  <ItemGroup>
+    <None Remove="Assets\Fonts\SourceSansPro-Bold.ttf" />
```
[the remark this answers is not in the document]
Comment (Gillibald) [reply samples/ControlCatalog/ControlCatalog.csproj]: I have no idea why visual studio added them. They are not needed.

**Review by grokys — APPROVED** [review]

**Comment by jpgarza93** [comment]
Can we add this to the docs? I am having trouble finding info on how to add a custom font.

**Comment by timunie** [comment]
@jpgarza93 if you now understood what to do, you are welcome to provide a PR to the docs: https://github.com/AvaloniaUI/Documentation

**Comment by jpgarza93** [comment]
@timunie I didn't understand :( but once I do I will, thanks
